tracker/src/main/java/org/matomo/sdk/extra/InstallReferrerReceiver.java

```diff
@@ -34,12 +33,13 @@ public void onReceive(Context context, Intent intent) {
             Timber.tag(TAG).d("Dropping forwarded intent");
             return;
         }
-        SharedPreferences preferences = Matomo.getInstance(context.getApplicationContext()).getPreferences();
         if (intent.getAction().equals(REFERRER_SOURCE_GPLAY)) {
             String referrer = intent.getStringExtra(ARG_KEY_GPLAY_REFERRER);
             if (referrer != null) {
-                preferences.edit().putString(PREF_KEY_INSTALL_REFERRER_EXTRAS, referrer).apply();
-                Timber.tag(TAG).d("Stored Google Play referrer extras: %s", referrer);
+                new Thread(() -> {
```

Comment:
Launching a thread is costly too.
The `apply` is already async, it changes an internal map, and IO happens offthread already. The cost for updating a map is negligible.

So the question becomes whether it is necessary that SharedPreferences are initialized on an extra thread. I have never found that be necessary. Is this something you have experienced?

Reply:
Yes I've seen quite a few times it and use this fix in prod since a few months.

The problem is not the map but how SharedPref works, when you initialize them a new thread is started to load the xml and every access is blocked via a lock. So until the preferences are loaded the edit and the putString will lock waiting for the load to be finished.
Load that can take any amount of time as disk based and OS queue can be full.

Starting a thread is way less costly than an main thread block :)

Anyway in all cases you might also want to read : https://android-developers.googleblog.com/2019/11/still-using-installbroadcast-switch-to.html as in a few month this will not work anymore at all and the other way will be mandatory.